dali/kernels/shape.h

```diff
@@ -0,0 +1,430 @@
+// &) { return 1; }
+
+// template <typename T, size_t N>
+// constexpr int ShapeDim(T (&)[N]) { return int(N); }
+
+// template <typename T>
+// constexpr int ShapeDim(const T &t) {
+//   return int(t.size());
+// }
+
+// /// @brief Calculates flat index of a given element in the tensor
+// /// @remarks If pos has fewer dimensions than shape, the remaining offsets are assumed to be 0
+// template <typename Shape, typename SamplePos>
+// ptrdiff_t CalcOffset(const Shape &shape, const SamplePos &pos) {
+//   ptrdiff_t ofs = pos[0];
+//   const int m = ShapeDim(pos);
+//   const int n = ShapeDim(shape);
+//   int i;
+//   for (i = 1; i < m; i++) {
+//     ofs *= shape[i];
+//     ofs += pos[i];
+//   }
+//   for (; i < n; i++) {
+//     ofs *= shape[i];
+//   }
+//   return ofs;
+// }
+
+/// @brief Returns the product of all elements in shape
+/// @param shape - shape of a tensor whose elements we count
+template <typename T>
+inline int64_t volume(const T &shape) {
+  int n = shape.size();
+  if (n < 1) {
+    return 0;
+  }
+  int64_t v = shape[0];
+  for (int i = 1; i < n; i++) {
+    v *= shape[i];
+  }
+  return v;
+}
+
+constexpr int DynamicDimensions = -1;
+
+template <int ndim>
+struct TensorShape;
+
+template <>
+struct TensorShape<DynamicDimensions> {
+  // TODO overkill like standard containers with all the iterator, elemet_types, reference_type etc?
+  // or just use decltype(shape.begin()) begin() { return shape.begin(); } ?
+  using element_type = int64_t;
+  using value_type = int64_t;
+  using size_type = int;
+  using reference = value_type &;
+  using const_reference = const value_type &;
+  using container_type = std::vector<element_type>;
+  using iterator = typename container_type::iterator;
+  using const_iterator = typename container_type::const_iterator;
+  using reverse_iterator = typename container_type::reverse_iterator;
+  using const_reverse_iterator = typename container_type::const_reverse_iterator;
+
+  TensorShape(const std::vector<int64_t> &s) : shape(s) {}
+
+  template <size_t N>
+  TensorShape(std::array<int64_t, N> s) : shape(s.begin(), s.end()) {}
+
+  template <typename... Ts>
+  TensorShape(int64_t i0, Ts... s) : shape{i0, int64_t{s}...} {}
+
+  TensorShape() = default;
+  TensorShape(const TensorShape &) = default;
+  TensorShape(TensorShape &&) = default;
+  TensorShape &operator=(const TensorShape &other) = default;
+  TensorShape &operator=(TensorShape &&other) = default;
+
+  template <int other_ndim>
+  TensorShape(const TensorShape<other_ndim> &other)
+      : shape(other.shape.begin(), other.shape.end()) {}
+
+  template <int other_ndim>
+  TensorShape &operator=(const TensorShape<other_ndim> &other) {
+    shape = std::vector<int64_t>(other.shape.begin(), other.shape.end());
+    return *this;
+  }
+
+  // We allow only explicit conversion from dynamic to static dim
+  template <int other_ndim>
+  TensorShape<other_ndim> to_static_ndim() const {
```

Comment:
Why not `to_static`? Again, too much typing for my liking...

Reply:
Done